File: Swift/DocEnumerator.swift
//
//  DocEnumerator.swift
//  SwiftForest
//
//  Created by Jens Alfke on 11/5/15.
//  Copyright (c) 2015-2016 Couchbase. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//  Unless required by applicable law or agreed to in writing, software distributed under the
//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
//  either express or implied. See the License for the specific language governing permissions
//  and limitations under the License.

import Foundation

/********
 NOTE: THIS IS A PROVISIONAL, PLACEHOLDER API, NOT THE OFFICIAL COUCHBASE LITE 2.0 API.
 It's for prototyping, experimentation, and performance testing. It will change without notice.
 Once the 2.0 API is designed, we will begin implementing that and remove these classes.
 ********/



/** Represents a range of docIDs in a Database, plus other query options.
    Acts as a Swift Sequence that generates a DocEnumerator for the actual enumeration.
    As a Sequence, this type can be directly used in a "for...in..." loop. */
public struct DocumentRange: LazySequenceProtocol {

    public init(database: Database) {
        self.database = database
    }

    public init(database: Database, docIDs: [String]) {
        self.database = database
        self.docIDs = docIDs
    }

    public let database: Database

    public var start: String? = nil
    public var end: String? = nil
    public var docIDs: [String]? = nil
    public var skip: UInt64 = 0
    public var limit: UInt64 = UInt64.max
    public var flags: EnumeratorFlags = [.InclusiveStart, .InclusiveEnd, .IncludeNonConflicted, .IncludeBodies]

    public func enumerateDocs() throws -> DocEnumerator {
        return try withC4Opts() { (c4opts, err: inout C4Error) in
            if docIDs == nil {
                return c4db_enumerateAllDocs(database.dbHandle, C4Slice(start), C4Slice(end), c4opts, &err)
            } else {
                var docIDSlices = docIDs!.map {C4Slice($0)}
                return c4db_enumerateSomeDocs(database.dbHandle, &docIDSlices, docIDSlices.count, c4opts, &err)
            }
        }
    }

    // common subroutine for creating enumerators; provides C4EnumeratorOptions, checks errors,
    // and wraps enumerator in a DocEnumerator
    fileprivate func withC4Opts(_ block: (UnsafePointer<C4EnumeratorOptions>, inout C4Error)->OpaquePointer) throws -> DocEnumerator {
        var result: OpaquePointer? = nil
        var err = C4Error()
        var c4opt = C4EnumeratorOptions(skip: skip, flags: C4EnumeratorFlags(rawValue: flags.rawValue))
        withUnsafePointer(to: &c4opt) { c4optPtr in
            result = block(c4optPtr, &err)
        }
        guard result != nil else {
            throw err
        }
        return DocEnumerator(database: database, c4enum: result, limit: limit)
    }

    // SEQUENCETYPE API:

    public typealias Iterator = DocEnumerator
    public typealias Subsequence = DocumentRange

    public func dropFirst(_ n: Int) -> DocumentRange {
        var r = self
        r.skip = UInt64(Int64(r.skip) + Int64(n))
        return r
    }

    func prefix(_ maxLength: Int) -> DocumentRange {
        var r = self
        r.limit = Swift.min(r.limit, UInt64(maxLength))
        return r
    }

    public func makeIterator() -> DocEnumerator {
        return try! enumerateDocs()     //FIX: Can't throw error
    }
}


open class DocEnumerator: IteratorProtocol, Swift.Sequence {

    public typealias Element = Document

    init(database: Database, c4enum: OpaquePointer?, limit: UInt64 = UInt64.max) {
        assert(c4enum != nil)
        self.database = database
        e = c4enum!
        self.limit = limit
    }

    deinit {
        c4enum_free(e)
    }

    open func next() -> Document? {
        guard limit > 0 else {
            return nil
        }
        limit -= 1

        do {
            return try nextDoc()
        } catch {
            return nil
        }
    }

    /** The same as next() except that it can throw an error. */
    open func nextDoc() throws -> Document? {
        var err = C4Error()
        let c4doc = c4enum_nextDocument(e, &err)
        guard c4doc != nil else {
            if err.code != 0 {
                self.error = err
                throw err
            }
            return nil
        }
        return Document(database: database, handle: c4doc!)
    }

    /** Check this after next() returns nil, to see if it stopped due to an error. */
    open var error: C4Error?

    /** Throws if next() stopped due to an error. */
    open func checkError() throws {
        if error != nil {
            throw error!
        }
    }

    fileprivate let database: Database
    fileprivate var e: OpaquePointer
    fileprivate var limit: UInt64
}


public struct EnumeratorFlags : OptionSet {
    public let rawValue: UInt16
    public init(rawValue: UInt16) { self.rawValue = rawValue }
    init(_ flag: C4EnumeratorFlags) { self.rawValue = flag.rawValue }

    public static let Descending        = EnumeratorFlags(C4EnumeratorFlags.descending)
    public static let InclusiveStart    = EnumeratorFlags(C4EnumeratorFlags.inclusiveStart)
    public static let InclusiveEnd      = EnumeratorFlags(C4EnumeratorFlags.inclusiveEnd)
    public static let IncludeDeleted    = EnumeratorFlags(C4EnumeratorFlags.includeDeleted)
    public static let IncludeNonConflicted = EnumeratorFlags(C4EnumeratorFlags.includeNonConflicted)
    public static let IncludeBodies     = EnumeratorFlags(C4EnumeratorFlags.includeBodies)
}

File: CBForest/Storage/DocEnumerator.cc
//
//  DocEnumerator.cc
//  CBNano
//
//  Created by Jens Alfke on 6/18/14.
//  Copyright (c) 2014 Couchbase. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//  Unless required by applicable law or agreed to in writing, software distributed under the
//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
//  either express or implied. See the License for the specific language governing permissions
//  and limitations under the License.

#include "DocEnumerator.hh"
#include "KeyStore.hh"
#include "LogInternal.hh"
#include <algorithm>
#include <limits.h>
#include <string.h>

using namespace std;


namespace cbforest {

#pragma mark - ENUMERATION:


    const DocEnumerator::Options DocEnumerator::Options::kDefault = {
        0,
        UINT_MAX,
        false,
        true,
        true,
        false,
        kDefaultContent,
    };


    DocEnumerator::DocEnumerator(KeyStore &store, const Options& options)
    :_store(&store),
     _options(options)
    { }


    // Key-range constructor
    DocEnumerator::DocEnumerator(KeyStore &store,
                                 slice startKey, slice endKey,
                                 const Options& options)
    :DocEnumerator(store, options)
    {
        Debug("enum: DocEnumerator(%s, [%s] -- [%s]%s) --> %p",
              store.name().c_str(), startKey.hexCString(), endKey.hexCString(),
              (options.descending ? " desc" : ""), this);
        if (startKey.size == 0)
            startKey.buf = NULL;
        if (endKey.size == 0)
            endKey.buf = NULL;

        slice minKey = startKey, maxKey = endKey;
        if (options.descending)
            swap(minKey, maxKey);

        _impl.reset(_store->newEnumeratorImpl(minKey, maxKey, _options));
        _skipStep = _impl->shouldSkipFirstStep();
    }

    // Sequence-range constructor
    DocEnumerator::DocEnumerator(KeyStore &store,
                                 sequence start, sequence end,
                                 const Options& options)
    :DocEnumerator(store, options)
    {
        Debug("enum: DocEnumerator(%s, #%llu -- #%llu) --> %p",
                store.name().c_str(), start, end, this);

        sequence minSeq = start, maxSeq = end;
        if (options.descending)
            swap(minSeq, maxSeq);

        _impl.reset(_store->newEnumeratorImpl(minSeq, maxSeq, _options));
        _skipStep = _impl->shouldSkipFirstStep();
    }

    // Key-array constructor
    DocEnumerator::DocEnumerator(KeyStore &store,
                                 vector<string> docIDs,
                                 const Options& options)
    :DocEnumerator(store, options)
    {
        _docIDs = docIDs;
        Debug("enum: DocEnumerator(%s, %zu keys) --> %p",
                store.name().c_str(), docIDs.size(), this);
        if (_options.skip > 0)
            _docIDs.erase(_docIDs.begin(), _docIDs.begin() + _options.skip);
        if (_options.limit < _docIDs.size())
            _docIDs.resize(_options.limit);
        if (_options.descending)
            reverse(_docIDs.begin(), _docIDs.end());
        // (this mode doesn't actually create an fdb_iterator)
    }

    DocEnumerator::~DocEnumerator() {
        //Debug("enum: ~DocEnumerator(%p)", this);
        close();
    }

    // Assignment from a temporary
    DocEnumerator& DocEnumerator::operator=(DocEnumerator&& e) {
        Debug("enum: operator= %p <-- %p", this, &e);
        _store = e._store;
        _impl = move(e._impl);
        e._impl = nullptr;
        _docIDs = e._docIDs;
        _curDocIndex = e._curDocIndex;
        _options = e._options;
        _skipStep = e._skipStep;
        return *this;
    }


    void DocEnumerator::close() {
        _doc.clear();
        _impl.reset();
    }


    bool DocEnumerator::next() {
        // Enumerating an array of docs is handled specially:
        if (_docIDs.size() > 0)
            return nextFromArray();

        if (!_impl)
            return false;
        if (_options.limit-- == 0) {
            close();
            return false;
        }
        do {
            if (_skipStep) {
                _skipStep = false;
            } else if (!_impl->next()) {
                close();
                return false;
            }
        } while (_options.skip > 0 && _options.skip-- > 0);
        return getDoc();
    }

    // implementation of next() when enumerating a vector of keys
    bool DocEnumerator::nextFromArray() {
        if (_curDocIndex >= _docIDs.size()) {
            Debug("enum: at end of vector");
            close();
            return false;
        }
        _doc.clearMetaAndBody();
        _doc.setKey(_docIDs[_curDocIndex++]);
        _store->read(_doc);
        Debug("enum:     fdb_get --> [%s]", _doc.key().hexCString());
        return true;
    }

    void DocEnumerator::seek(slice key) {
        Debug("enum: seek([%s])", key.hexCString());
        if (!_impl)
            return;

        _doc.clear();
        if (_impl->seek(key))
            _skipStep = true; // so next() won't skip over the doc
        else
            close();
    }

    bool DocEnumerator::getDoc() {
        _doc.clear();
        if (!_impl->read(_doc)) {
            close();
            return false;
        }
        Debug("enum:     fdb_iterator_get --> [%s]", _doc.key().hexCString());
        return true;
    }

}
